Update/enable frontend upgrade nudges

stacimc opened the pull request:
Fixes 158-gh-Automattic/view-design and 157-gh-Automattic/view-design and 175-gh-Automattic/view-design.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* re-enables frontend previews for paid blocks requiring an upgrade. If a block is unavailable due to not meeting plan requirements _and the current user is an admin_, the block will still be rendered on the frontend with an upgrade banner identical to the one that appears in the editor.
* refactors the existing UpgradePlanBanner used in the editor to pull out as much common code as possible into a generic Nudge component that can be server-side-rendered
* adds a frontend Stripe nudge for the Premium Block specifically, unless the upgrade nudge is already being displayed.

Upgrade nudge
<img width="789" alt="Screen Shot 2021-01-07 at 7 22 40 PM" src="https://user-images.githubusercontent.com/63313398/103970716-c2db8580-511d-11eb-8ad0-29634f26f27a.png">

Stripe nudge
<img width="654" alt="Screen Shot 2021-01-07 at 7 05 33 PM" src="https://user-images.githubusercontent.com/63313398/103970398-0aaddd00-511d-11eb-9b7b-55636201a854.png">

**On Atomic sites, the Stripe nudge redirects the user back to the post in the editor in order to edit the block/connect Stripe, rather than connecting directly to Stripe checkout**

#### Jetpack product discussion

Internal reference: pcjTuq-8U-p2

#### Does this pull request change what data or activity we track or use?

* TBD

#### Testing instructions:

This requires re-testing across all the scenarios: wpcom and Jetpack, with and without the required plan, and with and without Stripe connected.

##### Test setup notes
* the wp-env scripts will not sync changes to `class.jetpack-gutenberg.php`, `components.php`, or `class-blocks.php`. These may need to be manually synced to their respective files in the sandbox.
* before syncing, re-run `npm run build` from within the Jetpack dir in order to build the SSR templates.

##### Test scenarios
Jetpack
* Simply verify that you cannot add a Premium Content block.

**Wpcom**
1. On a wpcom site with a free plan, add a Premium Content block.
2. Preview the block in the frontend and verify that the block is displayed with an Upgrade nudge (no Stripe nudge).
3. View the post in an incognito window and verify that you cannot see the block.
4. Upgrade the site to a Personal plan.
5. View the post on the frontend again and verify that you now see the Stripe nudge.
6. Open the post in an incognito window and verify that you cannot see the block or the nudge.
7. Go back to the editor and connect Stripe.
8. Preview the block and verify that it is displayed with no nudges.
9. Revisit the post in an incognito window and verify that you can now see the block.

**Atomic (need to use business site so Stripe access works)**
* Create a WordPress.com business site and install a plugin so it becomes an Atomic site.
* Access the site via FTP
* Upload the Jetpack Beta plugin
* Delete the symlink to "Full Site Editing" in plugins
* Via wp-admin search for an install "Editing Toolkit"
* Activate both the Jetpack beta and editing toolkit plugins
* Via FTP edit wp-content/plugins/full-site-editing/full-site-editing.php and comment out line 304.
* Select this branch in the Jetpack beta plugin.
* Follow the same steps as for wpcom (minus free site), but note that the Stripe nudge's CTA should take you back to the editor.

Other Paid Blocks (Wpcom)
__(Pay with Paypal, Donations, ~~Payments,~~ Video, WhatsApp, Open Table, Calendly)__
1. On a wpcom site with a free plan, add the paid block
2. Preview the block in the frontend and verify that it is displayed with an Upgrade nudge
3. View the post in an incognito window and verify that you cannot see the block.
4. Upgrade the site to the necessary plan level for that block
5. View the post on the frontend and verify that you can see the block with no upgrade nudge
6. View the post in an incognito window and verify that you can now see the block.


#### Proposed changelog entry for your changes:

* TBD

stacimc committed:
Enable frontend preview for all paid blocks when upgrade required

When a paid block is inserted on a site that does not meet its
plan requirements, it will still be displayed on the frontend for
admin users, along with an upgrade nudge.

stacimc committed:
Update nudge to use the same styles as editor nudge

Updates the SSR rendered upgrade nudge for the frontend to use the
same styling as the upgrade nudge that displays in the editor.
Also does some refactoring in order to allow for some reuse of
existing components.

stacimc committed:
Display the Premium Content block in preview

The Premium Content block was not rendering in its preview because
its own pre-render checks were failing. This commit updates the
pre-render checks to pass when the user can edit the post.

A follow-up PR will enable the stripe nudge for this block when
the membership checks fail.

stacimc committed:
Refactor frontend nudge to allow configurable text

Small refactor to allow reuse of the SSR rendered frontend nudge
for other nudges, by making the description and button text
configurable. This will be used by Premium Content for a Stripe
nudge.

stacimc committed:
Linting

stacimc committed:
Render preset text for upgrade nudges

stacimc committed:
Linting

stacimc committed:
Render Stripe nudge on frontend for Premium Content block (#18245)

* Render Stripe nudge on frontend for Premium Content block

Renders a Stripe nudge on the frontend for admin users, if the
upgrade nudge is not already being displayed.

On wpcom sites the nudge directly links to the URL for connecting
to Stripe. On Jetpack sites it redirects to the post in the
editor, as the Stripe URL is not available from within the
render_callback in Jetpack.

* Use correct snake case

* Do not render block or nudge on Jetpack sites

The block will not be available on Jetpack sites so we no longer
need to handle this.

stacimc committed:
Linting

stacimc committed:
Redirect to editor for Premium Content stripe CTA on Atomic

The code which builds the Stripe connection url is not accessible
on Atomic sites, so in the Premium Content Stripe nudge, link the
user back to the post in the editor when the site is atomic.

stacimc commented on extensions/blocks/premium-content/premium-content.php:
```diff
@@ -64,6 +64,66 @@ function render_block( $attributes, $content ) {
+			get_edit_post_link( get_the_ID() ),
+			__( 'Connect to Stripe to use this block on your site.', 'jetpack' ),
+			__( 'Edit post', 'jetpack' )
+		);
```
I would love input on copy changes for the Stripe nudge on Atomic sites, to make it clearer to the user that they will be taken back to the editor.

apeatling replied:
"Connect to Stripe in the editor to use this block on your site." ?

stacimc replied:
Concise, +1. Thanks, copy is hard 😅

stacimc committed:
Update copy

apeatling commented:
Testing Instructions for Atomic (so Stripe connecting works):

* Create a WordPress.com business site and install a plugin so it becomes an Atomic site.
* Access the site via FTP
* Upload the Jetpack Beta plugin
* Delete the symlink to "Full Site Editing" in plugins
* Via wp-admin search for an install "Editing Toolkit"
* Activate both the Jetpack beta and editing toolkit plugins
* Via FTP edit `wp-content/plugins/full-site-editing/full-site-editing.php` and comment out line 304.
* Select this branch in the Jetpack beta plugin.
* Test.

apeatling commented:
I've gone through a simple site, atomic site, free and paid. I've checked all existing paid blocks. Nice work!

This all worked very well except for one thing -- the Payments block seems to not display the notice on the front end. It also shows up in an icognito window, but that's also the case on master.

apeatling reviewed (dismissed):
This looks good, the only thing I see is to clean up the missing period marks at the end of comments that the linter is complaining about.

stacimc commented:
>  the Payments block seems to not display the notice on the front end. It also shows up in an incognito window, but that's also the case on master.

@apeatling Thanks for pointing this out. Do you mean the Recurring Payments block? That block is unfortunately different because it does not use the `plan_check` attribute (all of its nudges are custom handled, in order to handle the `jetpack_block_editor_enable_upgrade_nudge` being false on jetpack). It will not get the frontend nudges for free.

I've crossed that block out in the list of paid blocks to test. Do you think it's worth a follow-up to add this to Payments for consistency? This is yet another annoying sticking point caused by that block's custom registration.

apeatling commented:
@stacimc Can you make an issue as a follow up? No need to work on it for this PR.

jeherve commented:
Do you think we could merge #17907 first, and then focus on this one? I think it would help, as the original Premium content block is getting a bit unwieldy.

stacimc commented:
> Do you think we could merge #17907 first, and then focus on this one? I think it would help, as the original Premium content block is getting a bit unwieldy.

@jeherve Totally agreed, we'd like this work to be considered a follow-up. Now that #17907 is merged I've rebased and updated the base branch to master.

jeherve committed:
Remove file now that it's error-free

jeherve committed:
Remove unused imports

jeherve committed:
Merge remote-tracking branch 'origin/master' into update/enable-frontend-upgrade-nudges

jeherve committed:
Remove unused variable

jeherve committed:
Avoid phpcs warnings for functions in global space

jeherve commented on projects/plugins/jetpack/extensions/shared/components/upgrade-nudge/style.scss:
```diff
@@ -1,13 +1,41 @@
+		margin-left: auto;
+		height: 28px; }
+		.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .components-button.is-primary {
+			background: #e34c84;
```
I wonder if we could switch to using studio colors for those colors; it may prove easier to maintain. Something like this:

```diff
diff --git a/projects/plugins/jetpack/extensions/shared/components/upgrade-nudge/style.scss b/projects/plugins/jetpack/extensions/shared/components/upgrade-nudge/style.scss
index d95940b84..921bf6872 100644
--- a/projects/plugins/jetpack/extensions/shared/components/upgrade-nudge/style.scss
+++ b/projects/plugins/jetpack/extensions/shared/components/upgrade-nudge/style.scss
@@ -1,4 +1,5 @@
 @import '../../styles/gutenberg-base-styles.scss';
+@import "~@automattic/color-studio/dist/color-variables";
 
 .jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper {
 	display: flex;
@@ -6,13 +7,13 @@
 	align-items: center;
 	font-size: 14px;
 	height: 48px;
-	background: black;
+	background: $studio-black;
 	padding: 0 20px;
 	border-radius: 2px;
-	box-shadow: 0 0 1px inset #fff; }
+	box-shadow: 0 0 1px inset $studio-white; }
 	.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .banner-title,
 	.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .banner-description {
-		color: #fff; }
+		color: $studio-white; }
 	.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .jetpack-upgrade-plan-banner__title,
 	.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .jetpack-upgrade-plan-banner__description {
 		margin-right: 10px; }
@@ -22,13 +23,13 @@
 		margin-left: auto;
 		height: 28px; }
 		.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .components-button.is-primary {
-			background: #e34c84;
-			color: #fff; }
+			background: $studio-pink-40;
+			color: $studio-white; }
 		.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .components-button.is-primary:hover {
-			background: #eb6594; }
+			background: $studio-pink-30; }
 		.jetpack-upgrade-plan-banner .jetpack-upgrade-plan-banner__wrapper .components-button.is-primary.is-busy {
 			background-size: 100px 100%;
-			background-image: linear-gradient(-45deg, #e34c84 28%, #ab235a 28%, #ab235a 72%, #e34c84 72%); }
+			background-image: linear-gradient(-45deg, $studio-pink-40 28%, $studio-pink-60 28%, $studio-pink-60 72%, $studio-pink-40 72%); }
 
 .jetpack-upgrade-plan-banner.block-editor-block-list__block {
 	margin-top: 0;

```

What do you think?

stacimc replied:
Completely agree 👍

jeherve reviewed (approved):
This looks good. I pushed a few changes, tested, this should be good to merge. I spotted some minor issues, but nothing blocking.

**On WordPress.com Simple**

I tested with D55853-code, in Chrome.

- On Free sites, I think we could do with some extra padding in the upgrade nudge for the Payments button inside the Premium Content block:

![image](https://user-images.githubusercontent.com/426388/105509366-b59dc980-5ccd-11eb-8fb1-aa929d6c5af7.png)

- Once I've paid for a plan and connected to Stripe, I did not see the Payment button on the frontend, only in the editor. The problem disappeared when I created a Payment button outside of the Premium Content block and assigned a plan to the button. I suppose that's because the subscription plan was saved.
- The same thing happened in Atomic.

**On Atomic**

I didn't spot any issues.

**On Jetpack**

I can't see the block when only using the production bundle. 👍

jeherve commented:
Going to merge this now. I'll let you handle the WordPress.com counterpart with the other related diffs later :)

Do you want me to create issues for the comments I left above?

stacimc commented:
> Do you want me to create issues for the comments I left above?

@jeherve If you don't mind that would be great; I'm not positive I understand exactly how to repro. Thank you!

jeherve commented:
> create issues for the comments I left above

I created #18521 and #18522 to track things.

jeherve commented:
r220427-wpcom
